src/quartile.rs

```diff
@@ -1,132 +1,21 @@
 
     // Safety: this expect is eliminated at compile time, as the compiler can
     // trivially verify that EFF_BUCKETS <= BUCKETS.
     let mut bucket_copy: [u32; EFF_BUCKETS] = bucket[..EFF_BUCKETS]
         .try_into()
         .expect("EFF_BUCKETS is bigger than BUCKETS");
 
-    let mut spl = 0;
-    let mut spr = 0;
-    let mut l = 0;
-    let mut r = end;
-    loop {
-        let ret = partition(&mut bucket_copy, l, r);
-        match ret.cmp(&p2) {
-            Ordering::Greater => {
-                r = ret - 1;
-                short_cut_right[spr] = ret;
-                spr += 1;
-            }
-            Ordering::Less => {
-                l = ret + 1;
-                short_cut_left[spl] = ret;
-                spl += 1;
-            }
-            Ordering::Equal => {
-                q2 = bucket_copy[p2];
-                break;
-            }
-        }
-    }
-
-    short_cut_left[spl] = p2 - 1;
-    short_cut_right[spr] = p2 + 1;
-
-    let mut l = 0;
-    for mut r in short_cut_left.iter().take(spl + 1).copied() {
-        match r.cmp(&p1) {
-            Ordering::Greater => {
-                loop {
-                    let ret = partition(&mut bucket_copy, l, r);
-                    match ret.cmp(&p1) {
-                        Ordering::Greater => r = ret - 1,
-                        Ordering::Less => l = ret + 1,
-                        Ordering::Equal => {
-                            q1 = bucket_copy[p1];
-                            break;
-                        }
-                    }
-                }
-                break;
-            }
-            Ordering::Less => {
-                l = r;
-            }
-            Ordering::Equal => {
-                q1 = bucket_copy[p1];
-                break;
-            }
-        }
-    }
-
-    let mut r = end;
-    for mut l in short_cut_right.iter().take(spr + 1).copied() {
-        match l.cmp(&p3) {
-            Ordering::Less => {
-                loop {
-                    let ret = partition(&mut bucket_copy, l, r);
-                    match ret.cmp(&p3) {
-                        Ordering::Greater => r = ret - 1,
-                        Ordering::Less => l = ret + 1,
-                        Ordering::Equal => {
-                            q3 = bucket_copy[p3];
-                            break;
-                        }
-                    }
-                }
-                break;
-            }
-            Ordering::Greater => {
-                r = l;
-            }
-            Ordering::Equal => {
-                q3 = bucket_copy[p3];
-                break;
-            }
-        }
-    }
+    // XXX: this code replaces the C++ version with a Rust core equivalent.
+    // See <https://github.com/vthib/tlsh/pull/17>.
+    let (lesser, &mut q2, greater) = bucket_copy.select_nth_unstable(p2);
+    let (_, &mut q1, _) = lesser.select_nth_unstable(p1);
+    let (_, &mut q3, _) = greater.select_nth_unstable(p3 - p2 - 1);
```

Comment:
`p3 - p2 - 1` could just be replaced by `p1` and p3 eliminated (renaming p1 to first_quartile and p2 to median would make it clearer). But it's not like this is not trivially done by the compiler so it doesn't matter